MathlibTest/Junk.lean

```diff
@@ -0,0 +1,360 @@
+import Mathlib
+
+/-!
+# Junk values in Lean and Mathlib
+
+This file aims to document and test the various junk values used to keep functions total. Each
+example is both an example application of the junk value, and a place to document and justify the
+choice (hopefully, in addition to the actual definition).
+
+This file can act as a reference guide for people who want to e.g. verify some theorem statement is
+what they want, but perhaps are not so intimately familiar with all of Mathlib's API otherwise.
+
+Most theorems should be provable by a simple calls to `simp`, `norm_num`, or `rfl`.
+-/
+
+/-! ## Truncating subtraction -/
+
+/-- Nat subtraction: When subtracting a natural number `x` from a smaller one `y`, the value is
+always zero. This is because we want to be able to also return a natural number, and so we can't
+return a negative number. -/
+example : (3 : ℕ) - (5 : ℕ) = 0 := by
+  simp
+
+/-- Because Nat subtraction truncates, it does not always agree with Int subtraction. -/
+example : ↑(3 - 5 : ℕ) ≠ (3 - 5 : ℤ) := by
+  simp
+
+/-- Other types with a lower bound also truncate. `PNat`, the strictly positive natural numbers
+(starting at 1), truncate at 1. Accordingly, they truncate even when subtracting a number from
+itself. -/
+example : (3 : ℕ+) - (5 : ℕ+) = 1 ∧ (10 : ℕ+) - (10 : ℕ+) = 1 := by
+  and_intros <;> rfl
+
+/-- `NNRat` truncates at 0. Here, `3 / 2 - 4 = 0`. -/
+example : (3 : NNRat) / 2 - (4 : NNRat) = 0 := by
+  decide +kernel
+
+/-- `NNReal` truncates at 0. Here, `2 - π = 0`. -/
+example : (2 : NNReal) - ⟨Real.pi, Real.pi_nonneg⟩ = 0 := by
+  rw [tsub_eq_zero_iff_le]
+  exact_mod_cast Real.two_le_pi
+
+/-- `ENNReal` truncates subtraction at 0. Here, `π - ∞ = 0`. -/
+example : ENNReal.ofNNReal ⟨Real.pi, Real.pi_nonneg⟩ - ⊤ = 0 := by
+  simp
```

Comment:
can you also please give an example that has abstracted variables like `x - y = 0` with `x < y`? Because in that case, `simp` wouldn't be enough. And you could showcase the associated lemma -- `tsub_eq...`